File: Algorithms/235_Lowest_Common_Ancestor_of_a_Binary_Search_Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from typing import *


class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        a1: List[TreeNode] = []
        a2: List[TreeNode] = []

        self.get_ancestors(root, p.val, a1)
        self.get_ancestors(root, q.val, a2)

        min_len = min(len(a1), len(a2))
        lca: [TreeNode, None] = None
        for i in range(0, min_len):
            if a1[i] == a2[i]:
                lca = a1[i]
            else:
                break

        return lca

    def get_ancestors(self, root: TreeNode, target: int, ancestors: List[TreeNode]) -> None:
        if root:
            ancestors.append(root)
            if target < root.val:
                self.get_ancestors(root.left, target, ancestors)
            elif target > root.val:
                self.get_ancestors(root.right, target, ancestors)

File: Algorithms/235_Lowest_Common_Ancestor_of_a_Binary_Search_Tree/solution.go
package main

func Min(a int, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}

/**
 * Definition for TreeNode.
 * type TreeNode struct {
 *     Val int
 *     Left *ListNode
 *     Right *ListNode
 * }
 */
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	a1, a2 := []*TreeNode{}, []*TreeNode{}
	getAncestors(root, p, &a1)
	getAncestors(root, q, &a2)

	minLen := Min(len(a1), len(a2))
	var lca *TreeNode
	for i := 0; i < minLen; i++ {
		if a1[i].Val == a2[i].Val {
			lca = a1[i]
		} else {
			break
		}
	}
	return lca
}

func getAncestors(root, target *TreeNode, ancestors *[]*TreeNode) {
	if root != nil {
		*ancestors = append(*ancestors, root)
		if target.Val < root.Val {
			getAncestors(root.Left, target, ancestors)
		} else if target.Val > root.Val {
			getAncestors(root.Right, target, ancestors)
		}
	}
}
